### PR TITLE
[DTensor] Optimize redistribute by merging allreduces on flattened meshes

### LINKED ISSUE #171916: DTensor must generate flattened PGs to avoid allreduce result inconsistency across Replicate when reducing over multiple mesh dims
### 🐛 Describe the bug

Suppose you have a mesh `(ddp=8, fsdp=5, tp=1)` (yes, you read that right, not (5, 8)) where your NVLink domain is size 8. This results in a jagged assignment of nodes across NVLink domains:

Image

Let's say you have a DTensor which is `(Partial, Partial, Replicate)` and you want to redistribute it to `(Replicate, Replicate, Replicate)`. We will do this in two allreduces (e.g., first on fsdp and then on ddp). In principle, it would be better to do it in one allreduce over fsdp+ddp, but our historical contract from DTensor is that we will never generate new flattened device mesh dims on the fly, as this has setup cost, and without the flattened ddp+fsdp mesh dim, we don't actually have the PGs that can do the all ranks all reduce.

However, there is a very big problem with doing it in two steps: you won't get the same result on all ranks. Specifically, NCCL doesn't guarantee reductions inside nvlink domains are bitwise identical to reductions across nvlink domains. When we do a reduction on FSDP first, there are only some FSDP groups that are entirely in one NVLink domain, and others that aren't. You end up with different results for the first reduction here.

Although it is nice to have DTensor not generate new PGs, we get incorrect results if we do not generate new PGs. Thus, this seems to decisively favor that DTensor *should* generate new PGs so that it can always arrange that all reductions happen in one go.

https://github.com/pytorch/pytorch/pull/171913 is a bandaid fix that gets rid of simple situations, but it only applies when the flattened mesh already exists. I concretely propose that we SHOULD generate the flattened mesh as well.

### Versions

main

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #172119

Fixes #171916

When redistributing a DTensor with multiple Partial placements of the same
reduce_op type (e.g., Partial("sum") on dims A and C), this change detects
if a flattened DeviceMesh exists that covers those dimensions and uses a
single allreduce instead of multiple separate ones.

Key changes:
- Add _FlattenedTransformInfo as duck-type compatible with _TransformInfo,
  allowing the redistribute loop to handle both uniformly via mesh override
- Add _get_flattened_mesh_by_layout() to query for existing flattened meshes
  using layout comparison rather than name-based lookup
- Add _optimize_transform_infos_for_flattened_reductions() to group same-type
  reductions and replace with flattened transforms
- Modify redistribute_local_tensor() to use mesh_to_use pattern so flattened
  transforms flow through the same code path as regular transforms

Example: For a (2,2,2) mesh with dims (A,B,C) and placements
(Partial("sum"), Partial("max"), Partial("sum")), if mesh["A","C"]._flatten()
was called, the A and C sums are merged into 1 allreduce (2 comms instead of 3).

Test Plan:
- Unit tests for _optimize_transform_infos_for_flattened_reductions using
  fake process group (fast, no NCCL init overhead)
- Integration tests verifying comm counts with CommDebugMode

-- Claude

### CODE REVIEW

File: torch/distributed/tensor/_redistribute.py [hunk 1]
```diff
@@ -107,6 +107,183 @@ class _TransformInfo(NamedTuple):
...
+                            if src_placements[dim].is_partial():
+                                # There's a Partial on a skipped dim - can't merge
+                                can_merge = False
+                                break
+
+                        if can_merge:
+                            same_type_reductions.append((j, next_info))
+                            last_mesh_dim = next_info.mesh_dim
+                            j += 1
+                        else:
```
Comment (ngimel) [inline torch/distributed/tensor/_redistribute.py]: you already broke if `can_merge=False`

File: torch/distributed/tensor/_redistribute.py [hunk 2]
```diff
@@ -107,6 +107,183 @@ class _TransformInfo(NamedTuple):
...
+                next_src, next_dst = next_info.src_dst_placements
+
+                if next_src.is_partial() and next_dst.is_replicate():
+                    if next_src == src:
+                        # Same type partial reduction - check if we can merge
+                        # We need to verify no Partial placements on skipped dims
+                        can_merge = True
+                        for dim in range(last_mesh_dim + 1, next_info.mesh_dim):
+                            if src_placements[dim].is_partial():
+                                # There's a Partial on a skipped dim - can't merge
```
Comment (ngimel) [inline torch/distributed/tensor/_redistribute.py]: why would a partial on a skipped dim prevent merge? PPP-> RPR can still be merged, no?

Comment (anshul-si) [reply torch/distributed/tensor/_redistribute.py]: this is true if the partial being skipped is of the same type + same_reduce_op as the source

Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py]: ah, good point. i think i can adjust this to allow skipping a same-type partial and only bail out if its a different type partial.

Updated to include this case.

File: torch/distributed/tensor/_redistribute.py (hunk 1 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py]: its confusing... but check the 'whole file' view. I think it's correct.

File: torch/distributed/tensor/_redistribute.py [hunk 3]
```diff
@@ -107,6 +107,186 @@ class _TransformInfo(NamedTuple):
...
+    flattened DeviceMesh exists.
+
+    IMPORTANT: Reductions can only be merged if there are no Partial placements
+    on the mesh dimensions between them. A Partial placement (even if unchanged)
+    implies a semantic ordering that must be preserved. For example:
+        (Psum, Pmax, Psum) -> the correct order is sum_A, then max_B, then sum_C
+    If we merge the sums: sum_{A,C}(x) then max_B gives a different result.
+
+    However, non-Partial placements (Replicate, Shard) in between are safe to
+    skip because they don't impose ordering constraints.
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py]: Does this mean if there are [Psum->R, S(0)->R, Psum->R] in transform_infos, we can reorder as [Psum->R, Psum->R, S(0)->R]? If so, this becomes we can do stable sort for `transform_infos` and place all Partial operations in the front or back.

File: torch/distributed/tensor/_redistribute.py [hunk 4]
```diff
@@ -107,6 +107,186 @@ class _TransformInfo(NamedTuple):
...
+            same_type_reductions = [(i, info)]
+            skipped_transforms: list[tuple[int, _TransformInfo]] = []
+            j = i + 1
+            last_mesh_dim = info.mesh_dim
+
+            while j < len(transform_infos):
+                next_info = transform_infos[j]
+                next_src, next_dst = next_info.src_dst_placements
+
+                if next_src.is_partial() and next_dst.is_replicate():
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py]: Suggest `if next_src == src and next_dst.is_replicate():`, we can reduce 7 LoC.

File: torch/distributed/tensor/_redistribute.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py]: Yes, if you think it's cleaner to write it that way I can change it to do an explicit sorting pass to group like-kind reductions, then do a second pass to merge adjacent like kind reductions.

File: torch/distributed/tensor/_redistribute.py (hunk 4 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py]: i got claude to make a more legible rewrite.

File: torch/distributed/tensor/_redistribute.py [hunk 5]
```diff
@@ -107,6 +107,157 @@ class _TransformInfo(NamedTuple):
...
+    expected_layout = submesh._layout.coalesce()
+    if len(expected_layout) > 1:
+        expected_layout = expected_layout.nest()
+
+    # Search existing flattened meshes by comparing layouts
+    for flattened_mesh in root_mesh._flatten_mapping.values():
+        if flattened_mesh._layout == expected_layout:
+            return flattened_mesh
+
+    return None
```
Comment (ngimel) [inline torch/distributed/tensor/_redistribute.py:170]: we should at the very least warn here and instruct users how to create flattened mesh. Ideally we'd error our (because not having a flattened mesh is a significant perf hit).

File: torch/distributed/tensor/_redistribute.py [hunk 6]
```diff
@@ -107,6 +107,157 @@ class _TransformInfo(NamedTuple):
...
+        while j < len(transform_infos):
+            next_info = transform_infos[j]
+            next_src, next_dst = next_info.src_dst_placements
+            is_reduction = next_src.is_partial() and next_dst.is_replicate()
+
+            if not is_reduction:
+                others.append((j, next_info))
+                j += 1
+            elif next_src == src and not has_blocking_partial(
+                src, reductions[-1][1].mesh_dim, next_info.mesh_dim
```
Comment (zpcore) [inline torch/distributed/tensor/_redistribute.py:242]: This is kind of weird, the `has_blocking_partial` is checking all mesh dims between A.mesh_dim and B.mesh_dim, which is not related to ordering in `transform_infos`.

Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:242]: um, maybe i misunderstood your point. but i think this is actually crucial.  See my other PR to see if it convinces you:
https://github.com/pytorch/pytorch/pull/172277

File: torch/distributed/tensor/_redistribute.py (hunk 5 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:170]: ok- i think i'll add a warn_once here. do you prefer a warn repetedly?

File: test/distributed/tensor/test_redistribute.py [hunk 7]
```diff
@@ -1342,6 +1347,502 @@ def test_shard_order_same_data_as_strided_shard(self):
...
+        rank = dist.get_rank()
+        if rank == 0 or rank == 7:
+            local_tensor = torch.ones(8, 8, device=self.device_type) * 10
+        else:
+            local_tensor = torch.zeros(8, 8, device=self.device_type)
+
+        dt = DTensor.from_local(
+            local_tensor,
+            mesh,
+            (Partial("sum"), Partial("max"), Partial("sum")),
```
Comment (zpcore) [inline test/distributed/tensor/test_redistribute.py:1893]: Feel like there maybe an issue with `has_blocking_partial`.  We may need device mesh with a higher dimension to test. I think we can use `LocalTensor`, though there is no concept of NCCL domain, but we can still test the tensor output just in case the partial sequence get messed up. 

With `LocalTensor`, we probably can mock `redistribute_local_tensor` to take the a long sequence of made up `transform_infos` and compare the redistributed tensor with merged partial `transform_infos`.

File: torch/distributed/tensor/_redistribute.py (hunk 6 above)
Comment (zpcore) [reply torch/distributed/tensor/_redistribute.py:242]: I see, the assumption is true if we follow the default left-to-right order. Do we want to consider the non default order or the graph based redistribution here?

File: torch/distributed/tensor/_redistribute.py (hunk 5 above)
Comment (ngimel) [reply torch/distributed/tensor/_redistribute.py:170]: Tbh I prefer noisy warning here, given that it's an easy remedy

File: test/distributed/tensor/test_redistribute.py (hunk 7 above)
Comment (zpcore) [reply test/distributed/tensor/test_redistribute.py:1893]: No need to worry. I think it's correct for this PR with the default non-graph based redistribution as in https://github.com/pytorch/pytorch/pull/172119#discussion_r2684141945. I can follow up and verify if we want to extend the support for arbitrary `transform_infos` sequences.

File: torch/distributed/tensor/_redistribute.py (hunk 6 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:242]: what do you mean about non-default order?  we have only defined a concept of non-default _shard_ ordering, but that ordering specifically applies to the shard placements it names.  I was concluding from that that Partials are always in fixed L to R order, which I further documented in the other PR.  If you disagree, lmk and we can clarify what the behavior should be for partials.

File: test/distributed/tensor/test_redistribute.py (hunk 7 above)
Comment (wconstab) [reply test/distributed/tensor/test_redistribute.py:1893]: I didn't understand these comments. maybe we can discuss offline to clarify.

File: torch/distributed/tensor/_redistribute.py (hunk 5 above)
Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:170]: ok, i put in a warning

File: torch/distributed/tensor/_redistribute.py (hunk 6 above)
Comment (zpcore) [reply torch/distributed/tensor/_redistribute.py:242]: Curious if the transforminfos is below:
```
mesh dim 0, P(sum)->R
mesh dim 2, P(sum)->R
mesh dim 1, P(avg)->R
```
Based on the code, we will not merge the first two `P(sum)->R`, because of mesh dim 1's `P(avg)->R`. Is this true?

Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:242]: I got it now. Yea I think that could be a problem- I'll think about it and add a test case for it

Comment (wconstab) [reply torch/distributed/tensor/_redistribute.py:242]: ok- discussed offline, this sequence of transform info should be banned as it is not valid to do redistribution of partial in arbitrary order. we should define a partial order that we always use and then this code is OK.

**Review by tianyu-l — CHANGES_REQUESTED** [review]
I wonder if this fix addresses my concern in https://github.com/pytorch/pytorch/pull/171913#discussion_r2670848002

In fact, I'm not sure if this is an issue to be solved at DTensor level at all. Do you think we can first detect and error out, before merging a solution which may not be sufficient?

I think it's more urgent to solve the problem for application API, e.g. `replicate` (DDP) + Sequence Parallel. cc @weifengpy

**Comment by wconstab** [comment]
@tianyu-l i made a doc to discuss the tradeoffs: https://docs.google.com/document/d/1hJsnodQmHfs1QosNgR39HZNiOOzfnZ6bnALqonDpcDs/edit?usp=sharing

tldr, i think this PR is at best a stopgap solution that does improve performance, from that angle i'm ok with landing it, but i don't love it.  I advocate for a bigger change to both (1) devicemesh/processgroup, and (2) redistribution planner in DTensor

**Comment by wconstab** [comment]
closing as replaced by https://github.com/pytorch/pytorch/pull/172610
